File: cpp_hybrid/test-problem-statement-multimodal.test.cpp
#include "/home/runner/work/elizaos-cpp/elizaos-cpp/SWEagent/tests/test-problem-statement-multimodal.test.h"

any mockedChildProcess = as<std::shared_ptr<jest::Mocked<child_process>>>(child_process);

void Main(void)
{
    jest->mock(std::string("child_process"));
    describe(std::string("SWEBenchMultimodalProblemStatement"), [=]() mutable
    {
        shared exampleImageUrl = std::string("https://upload.wikimedia.org/wikipedia/commons/thumb/4/4c/Candide1759.jpg/330px-Candide1759.jpg");
        beforeEach([=]() mutable
        {
            jest->clearAllMocks();
        }
        );
        describe(std::string("Initialization"), [=]() mutable
        {
            it(std::string("should initialize with basic properties"), [=]() mutable
            {
                auto problemStatement = std::make_shared<SWEBenchMultimodalProblemStatement>(object{
                    object::pair{std::string("text"), std::string("Test problem statement")}, 
                    object::pair{std::string("issueImages"), array<string>{ exampleImageUrl }}, 
                    object::pair{std::string("id"), std::string("test_id")}
                });
                expect(problemStatement->text)->toBe(std::string("Test problem statement"));
                expect(problemStatement->issueImages)->toEqual(array<string>{ exampleImageUrl });
                expect(problemStatement->id)->toBe(std::string("test_id"));
                expect(problemStatement->type)->toBe(std::string("swe_bench_multimodal"));
            }
            );
            it(std::string("should work with empty images array"), [=]() mutable
            {
                auto problemStatement = std::make_shared<SWEBenchMultimodalProblemStatement>(object{
                    object::pair{std::string("text"), std::string("Test problem statement")}, 
                    object::pair{std::string("issueImages"), array<any>()}
                });
                expect(problemStatement->issueImages)->toEqual(array<any>());
            }
            );
        }
        );
        describe(std::string("Getting problem statement"), [=]() mutable
        {
            it(std::string("should return text when no images present"), [=]() mutable
            {
                auto problemStatement = std::make_shared<SWEBenchMultimodalProblemStatement>(object{
                    object::pair{std::string("text"), std::string("Test problem statement")}, 
                    object::pair{std::string("issueImages"), array<any>()}
                });
                auto result = problemStatement->getProblemStatement();
                expect(result)->toBe(std::string("Test problem statement"));
            }
            );
            it(std::string("should handle valid image with successful download"), [=]() mutable
            {
                mockedChildProcess->execSync->mockImplementationOnce(as<any>(([=](auto command, auto _options = undefined) mutable
                {
                    if (AND((type_of(command) == std::string("string")), (command->includes(std::string("-I"))))) {
                        return std::string("HTTP/1.1 200 OK
\
Content-Type: image/png
\
Content-Length: 1234
\

\
");
                    }
                    return string_empty;
                }
                )));
                mockedChildProcess->execSync->mockImplementationOnce(as<any>(([=](auto command, auto _options = undefined) mutable
                {
                    if (AND((type_of(command) == std::string("string")), (command->includes(std::string("base64"))))) {
                        return std::string("ZmFrZV9pbWFnZV9kYXRh");
                    }
                    return string_empty;
                }
                )));
                auto problemStatement = std::make_shared<SWEBenchMultimodalProblemStatement>(object{
                    object::pair{std::string("text"), std::string("Test problem statement")}, 
                    object::pair{std::string("issueImages"), array<string>{ exampleImageUrl }}
                });
                auto result = problemStatement->getProblemStatement();
                expect(result)->toContain(std::string("Test problem statement"));
                expect(result)->toContain(std::string("![") + exampleImageUrl + std::string("](data:image/png;base64,"));
                expect(mockedChildProcess->execSync)->toHaveBeenCalledTimes(2);
            }
            );
            it(std::string("should handle network errors gracefully"), [=]() mutable
            {
                mockedChildProcess->execSync->mockImplementationOnce(as<any>(([=]() mutable
                {
                    throw any(std::make_shared<Error>(std::string("Network error")));
                }
                )));
                auto problemStatement = std::make_shared<SWEBenchMultimodalProblemStatement>(object{
                    object::pair{std::string("text"), std::string("Test problem statement")}, 
                    object::pair{std::string("issueImages"), array<string>{ exampleImageUrl }}
                });
                auto result = problemStatement->getProblemStatement();
                expect(result)->toBe(std::string("Test problem statement"));
                expect(mockedChildProcess->execSync)->toHaveBeenCalled();
            }
            );
            it(std::string("should reject invalid MIME types"), [=]() mutable
            {
                mockedChildProcess->execSync->mockImplementationOnce(as<any>(([=](auto command, auto _options = undefined) mutable
                {
                    if (AND((type_of(command) == std::string("string")), (command->includes(std::string("-I"))))) {
                        return std::string("HTTP/1.1 200 OK
\
Content-Type: text/html
\
Content-Length: 1234
\

\
");
                    }
                    return string_empty;
                }
                )));
                auto problemStatement = std::make_shared<SWEBenchMultimodalProblemStatement>(object{
                    object::pair{std::string("text"), std::string("Test problem statement")}, 
                    object::pair{std::string("issueImages"), array<string>{ std::string("http://example.com/document.html") }}
                });
                auto result = problemStatement->getProblemStatement();
                expect(result)->toBe(std::string("Test problem statement"));
            }
            );
            it(std::string("should cache results and not re-download images"), [=]() mutable
            {
                mockedChildProcess->execSync->mockImplementationOnce(as<any>(([=](auto command, auto _options = undefined) mutable
                {
                    if (AND((type_of(command) == std::string("string")), (command->includes(std::string("-I"))))) {
                        return std::string("HTTP/1.1 200 OK
\
Content-Type: image/png
\
Content-Length: 1234
\

\
");
                    }
                    return string_empty;
                }
                )));
                mockedChildProcess->execSync->mockImplementationOnce(as<any>(([=](auto command, auto _options = undefined) mutable
                {
                    if (AND((type_of(command) == std::string("string")), (command->includes(std::string("base64"))))) {
                        return std::string("ZmFrZV9pbWFnZV9kYXRh");
                    }
                    return string_empty;
                }
                )));
                auto problemStatement = std::make_shared<SWEBenchMultimodalProblemStatement>(object{
                    object::pair{std::string("text"), std::string("Test problem statement")}, 
                    object::pair{std::string("issueImages"), array<string>{ exampleImageUrl }}
                });
                auto result1 = problemStatement->getProblemStatement();
                auto result2 = problemStatement->getProblemStatement();
                expect(mockedChildProcess->execSync)->toHaveBeenCalledTimes(2);
                expect(result1)->toBe(result2);
                expect(result1)->toContain(std::string("Test problem statement"));
                expect(result1)->toContain(std::string("![") + exampleImageUrl + std::string("](data:image/png;base64,"));
            }
            );
            it(std::string("should handle invalid URLs gracefully"), [=]() mutable
            {
                auto problemStatement = std::make_shared<SWEBenchMultimodalProblemStatement>(object{
                    object::pair{std::string("text"), std::string("Test problem statement")}, 
                    object::pair{std::string("issueImages"), array<string>{ std::string("not_a_url"), std::string("ftp://invalid_scheme.com/image.png") }}
                });
                auto result = problemStatement->getProblemStatement();
                expect(result)->toBe(std::string("Test problem statement"));
                expect(mockedChildProcess->execSync)->not->toHaveBeenCalled();
            }
            );
            it(std::string("should reject large images"), [=]() mutable
            {
                mockedChildProcess->execSync->mockImplementationOnce(as<any>(([=](auto command, auto _options = undefined) mutable
                {
                    if (AND((type_of(command) == std::string("string")), (command->includes(std::string("-I"))))) {
                        return std::string("HTTP/1.1 200 OK
\
Content-Type: image/png
\
Content-Length: 20971520
\

\
");
                    }
                    return string_empty;
                }
                )))->mockImplementationOnce(as<any>(([=](auto command, auto _options = undefined) mutable
                {
                    if (AND((type_of(command) == std::string("string")), (command->includes(std::string("base64"))))) {
                        throw any(std::make_shared<Error>(std::string("curl: (63) Maximum file size exceeded")));
                    }
                    return string_empty;
                }
                )));
                auto problemStatement = std::make_shared<SWEBenchMultimodalProblemStatement>(object{
                    object::pair{std::string("text"), std::string("Test problem statement")}, 
                    object::pair{std::string("issueImages"), array<string>{ std::string("http://example.com/huge_image.png") }}
                });
                auto result = problemStatement->getProblemStatement();
                expect(result)->toBe(std::string("Test problem statement"));
                expect(mockedChildProcess->execSync)->toHaveBeenCalledTimes(2);
            }
            );
            it(std::string("should handle multiple images"), [=]() mutable
            {
                mockedChildProcess->execSync->mockImplementationOnce(as<any>(([=](auto command, auto _options = undefined) mutable
                {
                    if (AND((AND((type_of(command) == std::string("string")), (command->includes(std::string("-I"))))), (command->includes(std::string("image1.png"))))) {
                        return std::string("HTTP/1.1 200 OK
\
Content-Type: image/png
\
Content-Length: 1234
\

\
");
                    }
                    return string_empty;
                }
                )))->mockImplementationOnce(as<any>(([=](auto command, auto _options = undefined) mutable
                {
                    if (AND((AND((type_of(command) == std::string("string")), (command->includes(std::string("base64"))))), (command->includes(std::string("image1.png"))))) {
                        return std::string("aW1hZ2UxX2RhdGE=");
                    }
                    return string_empty;
                }
                )))->mockImplementationOnce(as<any>(([=](auto command, auto _options = undefined) mutable
                {
                    if (AND((AND((type_of(command) == std::string("string")), (command->includes(std::string("-I"))))), (command->includes(std::string("image2.jpg"))))) {
                        return std::string("HTTP/1.1 200 OK
\
Content-Type: image/jpeg
\
Content-Length: 1234
\

\
");
                    }
                    return string_empty;
                }
                )))->mockImplementationOnce(as<any>(([=](auto command, auto _options = undefined) mutable
                {
                    if (AND((AND((type_of(command) == std::string("string")), (command->includes(std::string("base64"))))), (command->includes(std::string("image2.jpg"))))) {
                        return std::string("aW1hZ2UyX2RhdGE=");
                    }
                    return string_empty;
                }
                )));
                auto problemStatement = std::make_shared<SWEBenchMultimodalProblemStatement>(object{
                    object::pair{std::string("text"), std::string("Test problem statement")}, 
                    object::pair{std::string("issueImages"), array<string>{ std::string("http://example.com/image1.png"), std::string("http://example.com/image2.jpg") }}
                });
                auto result = problemStatement->getProblemStatement();
                expect(result)->toContain(std::string("Test problem statement"));
                expect(result)->toContain(std::string("![http://example.com/image1.png](data:image/png;base64,"));
                expect(result)->toContain(std::string("![http://example.com/image2.jpg](data:image/jpeg;base64,"));
                expect(mockedChildProcess->execSync)->toHaveBeenCalledTimes(4);
            }
            );
            it(std::string("should handle mixed valid and invalid images"), [=]() mutable
            {
                mockedChildProcess->execSync->mockImplementationOnce(as<any>(([=](auto command, auto _options = undefined) mutable
                {
                    if (AND((AND((type_of(command) == std::string("string")), (command->includes(std::string("-I"))))), (command->includes(std::string("valid.png"))))) {
                        return std::string("HTTP/1.1 200 OK
\
Content-Type: image/png
\
Content-Length: 1234
\

\
");
                    }
                    return string_empty;
                }
                )))->mockImplementationOnce(as<any>(([=](auto command, auto _options = undefined) mutable
                {
                    if (AND((AND((type_of(command) == std::string("string")), (command->includes(std::string("base64"))))), (command->includes(std::string("valid.png"))))) {
                        return std::string("dmFsaWRfaW1hZ2U=");
                    }
                    return string_empty;
                }
                )))->mockImplementationOnce(as<any>(([=](auto _command, auto _options = undefined) mutable
                {
                    throw any(std::make_shared<Error>(std::string("Failed to load")));
                }
                )));
                auto problemStatement = std::make_shared<SWEBenchMultimodalProblemStatement>(object{
                    object::pair{std::string("text"), std::string("Test problem statement")}, 
                    object::pair{std::string("issueImages"), array<string>{ std::string("http://example.com/valid.png"), std::string("http://example.com/invalid.png") }}
                });
                auto result = problemStatement->getProblemStatement();
                expect(result)->toContain(std::string("Test problem statement"));
                expect(result)->toContain(std::string("![http://example.com/valid.png](data:image/png;base64,"));
                expect(result)->not->toContain(std::string("http://example.com/invalid.png"));
                expect(mockedChildProcess->execSync)->toHaveBeenCalledTimes(3);
            }
            );
            it(std::string("should handle HTTP error status codes"), [=]() mutable
            {
                mockedChildProcess->execSync->mockImplementationOnce(as<any>(([=](auto command, auto _options = undefined) mutable
                {
                    if (AND((type_of(command) == std::string("string")), (command->includes(std::string("-I"))))) {
                        throw any(std::make_shared<Error>(std::string("curl: (22) The requested URL returned error: 404 Not Found")));
                    }
                    return string_empty;
                }
                )));
                auto problemStatement = std::make_shared<SWEBenchMultimodalProblemStatement>(object{
                    object::pair{std::string("text"), std::string("Test problem statement")}, 
                    object::pair{std::string("issueImages"), array<string>{ std::string("http://example.com/missing.png") }}
                });
                auto result = problemStatement->getProblemStatement();
                expect(result)->toBe(std::string("Test problem statement"));
            }
            );
            it(std::string("should validate image URLs before attempting download"), [=]() mutable
            {
                auto problemStatement = std::make_shared<SWEBenchMultimodalProblemStatement>(object{
                    object::pair{std::string("text"), std::string("Test problem statement")}, 
                    object::pair{std::string("issueImages"), array<string>{ string_empty, std::string("javascript:alert(1)"), std::string("data:image/png;base64,abc"), std::string("file:///etc/passwd") }}
                });
                auto result = problemStatement->getProblemStatement();
                expect(mockedChildProcess->execSync)->not->toHaveBeenCalled();
                expect(result)->toBe(std::string("Test problem statement"));
            }
            );
        }
        );
    }
    );
}

MAIN

File: SWEagent/tests/test-problem-statement-multimodal.test.ts
/**
 * Multimodal problem statement tests converted from test_problem_statement_multimodal.py
 */

import { describe, it, expect, beforeEach, jest } from '@jest/globals';
import { SWEBenchMultimodalProblemStatement } from '../src/agent/problem-statement';
import * as child_process from 'child_process';

// Mock child_process.execSync
jest.mock('child_process');
const mockedChildProcess = child_process as jest.Mocked<typeof child_process>;

describe('SWEBenchMultimodalProblemStatement', () => {
  const exampleImageUrl =
    'https://upload.wikimedia.org/wikipedia/commons/thumb/4/4c/Candide1759.jpg/330px-Candide1759.jpg';

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('Initialization', () => {
    it('should initialize with basic properties', () => {
      const problemStatement = new SWEBenchMultimodalProblemStatement({
        text: 'Test problem statement',
        issueImages: [exampleImageUrl],
        id: 'test_id',
      });

      expect(problemStatement.text).toBe('Test problem statement');
      expect(problemStatement.issueImages).toEqual([exampleImageUrl]);
      expect(problemStatement.id).toBe('test_id');
      expect(problemStatement.type).toBe('swe_bench_multimodal');
    });

    it('should work with empty images array', () => {
      const problemStatement = new SWEBenchMultimodalProblemStatement({
        text: 'Test problem statement',
        issueImages: [],
      });

      expect(problemStatement.issueImages).toEqual([]);
    });
  });

  describe('Getting problem statement', () => {
    it('should return text when no images present', () => {
      const problemStatement = new SWEBenchMultimodalProblemStatement({
        text: 'Test problem statement',
        issueImages: [],
      });

      const result = problemStatement.getProblemStatement();
      expect(result).toBe('Test problem statement');
    });

    it('should handle valid image with successful download', () => {
      // Mock successful curl HEAD request for headers
      mockedChildProcess.execSync.mockImplementationOnce(((command: any, _options?: any) => {
        if (typeof command === 'string' && command.includes('-I')) {
          // HEAD request for headers
          return 'HTTP/1.1 200 OK\r\nContent-Type: image/png\r\nContent-Length: 1234\r\n\r\n';
        }
        return '';
      }) as any);

      // Mock successful curl download and base64 encoding
      mockedChildProcess.execSync.mockImplementationOnce(((command: any, _options?: any) => {
        if (typeof command === 'string' && command.includes('base64')) {
          // Download and base64 encode
          return 'ZmFrZV9pbWFnZV9kYXRh'; // base64 of 'fake_image_data'
        }
        return '';
      }) as any);

      const problemStatement = new SWEBenchMultimodalProblemStatement({
        text: 'Test problem statement',
        issueImages: [exampleImageUrl],
      });

      const result = problemStatement.getProblemStatement();

      expect(result).toContain('Test problem statement');
      expect(result).toContain(`![${exampleImageUrl}](data:image/png;base64,`);
      expect(mockedChildProcess.execSync).toHaveBeenCalledTimes(2);
    });

    it('should handle network errors gracefully', () => {
      // Mock network error
      mockedChildProcess.execSync.mockImplementationOnce((() => {
        throw new Error('Network error');
      }) as any);

      const problemStatement = new SWEBenchMultimodalProblemStatement({
        text: 'Test problem statement',
        issueImages: [exampleImageUrl],
      });

      const result = problemStatement.getProblemStatement();

      // Should return just the text, without throwing
      expect(result).toBe('Test problem statement');
      expect(mockedChildProcess.execSync).toHaveBeenCalled();
    });

    it('should reject invalid MIME types', () => {
      // Mock response with invalid MIME type
      mockedChildProcess.execSync.mockImplementationOnce(((command: any, _options?: any) => {
        if (typeof command === 'string' && command.includes('-I')) {
          // HEAD request with HTML content type
          return 'HTTP/1.1 200 OK\r\nContent-Type: text/html\r\nContent-Length: 1234\r\n\r\n';
        }
        return '';
      }) as any);

      const problemStatement = new SWEBenchMultimodalProblemStatement({
        text: 'Test problem statement',
        issueImages: ['http://example.com/document.html'],
      });

      const result = problemStatement.getProblemStatement();

      // Should return just the text
      expect(result).toBe('Test problem statement');
    });

    it('should cache results and not re-download images', () => {
      // Mock successful curl HEAD request for headers
      mockedChildProcess.execSync.mockImplementationOnce(((command: any, _options?: any) => {
        if (typeof command === 'string' && command.includes('-I')) {
          // HEAD request for headers
          return 'HTTP/1.1 200 OK\r\nContent-Type: image/png\r\nContent-Length: 1234\r\n\r\n';
        }
        return '';
      }) as any);

      // Mock successful curl download and base64 encoding
      mockedChildProcess.execSync.mockImplementationOnce(((command: any, _options?: any) => {
        if (typeof command === 'string' && command.includes('base64')) {
          // Download and base64 encode
          return 'ZmFrZV9pbWFnZV9kYXRh'; // base64 of 'fake_image_data'
        }
        return '';
      }) as any);

      const problemStatement = new SWEBenchMultimodalProblemStatement({
        text: 'Test problem statement',
        issueImages: [exampleImageUrl],
      });

      // Call getProblemStatement twice
      const result1 = problemStatement.getProblemStatement();
      const result2 = problemStatement.getProblemStatement();

      // Should only call execSync twice (once for HEAD, once for download) due to caching
      expect(mockedChildProcess.execSync).toHaveBeenCalledTimes(2);
      expect(result1).toBe(result2);
      expect(result1).toContain('Test problem statement');
      expect(result1).toContain(`![${exampleImageUrl}](data:image/png;base64,`);
    });

    it('should handle invalid URLs gracefully', () => {
      const problemStatement = new SWEBenchMultimodalProblemStatement({
        text: 'Test problem statement',
        issueImages: ['not_a_url', 'ftp://invalid_scheme.com/image.png'],
      });

      const result = problemStatement.getProblemStatement();

      // Should return just the text
      expect(result).toBe('Test problem statement');
      expect(mockedChildProcess.execSync).not.toHaveBeenCalled();
    });

    it('should reject large images', () => {
      // Mock response with large content-length header
      mockedChildProcess.execSync
        .mockImplementationOnce(((command: any, _options?: any) => {
          if (typeof command === 'string' && command.includes('-I')) {
            // HEAD request with large content-length
            return 'HTTP/1.1 200 OK\r\nContent-Type: image/png\r\nContent-Length: 20971520\r\n\r\n'; // 20MB
          }
          return '';
        }) as any)
        .mockImplementationOnce(((command: any, _options?: any) => {
          if (typeof command === 'string' && command.includes('base64')) {
            // Simulate curl failing due to max-filesize limit
            throw new Error('curl: (63) Maximum file size exceeded');
          }
          return '';
        }) as any);

      const problemStatement = new SWEBenchMultimodalProblemStatement({
        text: 'Test problem statement',
        issueImages: ['http://example.com/huge_image.png'],
      });

      const result = problemStatement.getProblemStatement();

      // Should return just the text, rejecting the large image
      expect(result).toBe('Test problem statement');
      expect(mockedChildProcess.execSync).toHaveBeenCalledTimes(2);
    });

    it('should handle multiple images', () => {
      // Mock responses for first image
      mockedChildProcess.execSync
        .mockImplementationOnce(((command: any, _options?: any) => {
          if (typeof command === 'string' && command.includes('-I') && command.includes('image1.png')) {
            return 'HTTP/1.1 200 OK\r\nContent-Type: image/png\r\nContent-Length: 1234\r\n\r\n';
          }
          return '';
        }) as any)
        .mockImplementationOnce(((command: any, _options?: any) => {
          if (typeof command === 'string' && command.includes('base64') && command.includes('image1.png')) {
            return 'aW1hZ2UxX2RhdGE='; // base64 of 'image1_data'
          }
          return '';
        }) as any)
        // Mock responses for second image
        .mockImplementationOnce(((command: any, _options?: any) => {
          if (typeof command === 'string' && command.includes('-I') && command.includes('image2.jpg')) {
            return 'HTTP/1.1 200 OK\r\nContent-Type: image/jpeg\r\nContent-Length: 1234\r\n\r\n';
          }
          return '';
        }) as any)
        .mockImplementationOnce(((command: any, _options?: any) => {
          if (typeof command === 'string' && command.includes('base64') && command.includes('image2.jpg')) {
            return 'aW1hZ2UyX2RhdGE='; // base64 of 'image2_data'
          }
          return '';
        }) as any);

      const problemStatement = new SWEBenchMultimodalProblemStatement({
        text: 'Test problem statement',
        issueImages: ['http://example.com/image1.png', 'http://example.com/image2.jpg'],
      });

      const result = problemStatement.getProblemStatement();

      expect(result).toContain('Test problem statement');
      expect(result).toContain('![http://example.com/image1.png](data:image/png;base64,');
      expect(result).toContain('![http://example.com/image2.jpg](data:image/jpeg;base64,');
      expect(mockedChildProcess.execSync).toHaveBeenCalledTimes(4); // 2 HEAD + 2 download requests
    });

    it('should handle mixed valid and invalid images', () => {
      // Mock successful response for first image
      mockedChildProcess.execSync
        .mockImplementationOnce(((command: any, _options?: any) => {
          if (typeof command === 'string' && command.includes('-I') && command.includes('valid.png')) {
            return 'HTTP/1.1 200 OK\r\nContent-Type: image/png\r\nContent-Length: 1234\r\n\r\n';
          }
          return '';
        }) as any)
        .mockImplementationOnce(((command: any, _options?: any) => {
          if (typeof command === 'string' && command.includes('base64') && command.includes('valid.png')) {
            return 'dmFsaWRfaW1hZ2U='; // base64 of 'valid_image'
          }
          return '';
        }) as any)
        // Mock failed response for second image
        .mockImplementationOnce(((_command: any, _options?: any) => {
          throw new Error('Failed to load');
        }) as any);

      const problemStatement = new SWEBenchMultimodalProblemStatement({
        text: 'Test problem statement',
        issueImages: ['http://example.com/valid.png', 'http://example.com/invalid.png'],
      });

      const result = problemStatement.getProblemStatement();

      expect(result).toContain('Test problem statement');
      expect(result).toContain('![http://example.com/valid.png](data:image/png;base64,');
      // Invalid image should be silently skipped
      expect(result).not.toContain('http://example.com/invalid.png');
      expect(mockedChildProcess.execSync).toHaveBeenCalledTimes(3); // 2 for valid image, 1 failed for invalid
    });

    it('should handle HTTP error status codes', () => {
      // Mock 404 response
      mockedChildProcess.execSync.mockImplementationOnce(((command: any, _options?: any) => {
        if (typeof command === 'string' && command.includes('-I')) {
          // Simulate 404 error
          throw new Error('curl: (22) The requested URL returned error: 404 Not Found');
        }
        return '';
      }) as any);

      const problemStatement = new SWEBenchMultimodalProblemStatement({
        text: 'Test problem statement',
        issueImages: ['http://example.com/missing.png'],
      });

      const result = problemStatement.getProblemStatement();

      // Should handle gracefully and return just the text
      expect(result).toBe('Test problem statement');
    });

    it('should validate image URLs before attempting download', () => {
      const problemStatement = new SWEBenchMultimodalProblemStatement({
        text: 'Test problem statement',
        issueImages: [
          '', // Empty string
          'javascript:alert(1)', // JavaScript URL
          'data:image/png;base64,abc', // Data URL (should be skipped)
          'file:///etc/passwd', // File URL
        ],
      });

      const result = problemStatement.getProblemStatement();

      // None of these should trigger HTTP requests
      expect(mockedChildProcess.execSync).not.toHaveBeenCalled();
      expect(result).toBe('Test problem statement');
    });
  });
});
